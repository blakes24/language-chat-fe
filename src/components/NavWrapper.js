import React, { useContext, useState } from "react";
import AppBar from "@material-ui/core/AppBar";
import CssBaseline from "@material-ui/core/CssBaseline";
import Divider from "@material-ui/core/Divider";
import Drawer from "@material-ui/core/Drawer";
import Hidden from "@material-ui/core/Hidden";
import IconButton from "@material-ui/core/IconButton";
import List from "@material-ui/core/List";
import ListItem from "@material-ui/core/ListItem";
import ListItemIcon from "@material-ui/core/ListItemIcon";
import ListItemText from "@material-ui/core/ListItemText";
import MailIcon from "@material-ui/icons/Mail";
import MenuIcon from "@material-ui/icons/Menu";
import Toolbar from "@material-ui/core/Toolbar";
import Typography from "@material-ui/core/Typography";
import { makeStyles, useTheme } from "@material-ui/core/styles";
import Link from "@material-ui/core/Link";
import { Link as RouterLink } from "react-router-dom";
import UserContext from "../helpers/UserContext";
import HomeIcon from "@material-ui/icons/Home";
import PeopleIcon from "@material-ui/icons/People";
import AccountCircleIcon from "@material-ui/icons/AccountCircle";
import ExitToAppIcon from "@material-ui/icons/ExitToApp";
import useScrollTrigger from "@material-ui/core/useScrollTrigger";
import Slide from "@material-ui/core/Slide";
import Footer from "./Footer";

const useStyles = makeStyles((theme) => ({
  root: {
    display: "flex",
  },
  drawer: {
    [theme.breakpoints.up("md")]: {
      width: 280,
      flexShrink: 0,
    },
  },
  appBar: {
    zIndex: theme.zIndex.drawer + 1,
  },
  menuButton: {
    marginRight: theme.spacing(2),
    [theme.breakpoints.up("md")]: {
      display: "none",
    },
  },
  // necessary for content to be below app bar
  toolbar: theme.mixins.toolbar,
  drawerPaper: {
    width: 220,
    [theme.breakpoints.up("md")]: {
      width: 280,
    },
    backgroundColor: theme.palette.secondary.light,
  },
  content: {
    flexGrow: 1,
    minHeight: "100vh",
  },
  link: {
    marginRight: theme.spacing(2),
  },
  title: {
    flexGrow: 1,
  },
  list: {
    margin: "1rem",
  },
  listItem: {
    marginTop: "1rem",
  },
}));

function HideOnScroll({ children }) {
  const trigger = useScrollTrigger();

  return (
    <Slide appear={false} direction="down" in={!trigger}>
      {children}
    </Slide>
  );
}

function NavWrapper({ window, children }) {
  const classes = useStyles();
  const theme = useTheme();
  const [mobileOpen, setMobileOpen] = useState(false);
  const { user, setToken } = useContext(UserContext);

  function logout() {
    setMobileOpen(false);
    setToken(null);
  }

  const handleDrawerToggle = () => {
    setMobileOpen(!mobileOpen);
  };

  const closeDrawer = () => {
    setMobileOpen(false);
  };

  const drawer = (
    <div>
      <div className={classes.toolbar} />
      <Divider />
      <List className={classes.list}>
        <ListItem
          button
          component={RouterLink}
          to="/"
          key={"home"}
          className={classes.listItem}
          onClick={closeDrawer}
        >
          <ListItemIcon>
            <HomeIcon color="secondary" />
          </ListItemIcon>
          <ListItemText primary={"Home"} />
        </ListItem>
        <ListItem
          button
          component={RouterLink}
          to="/partners"
          key={"partners"}
          className={classes.listItem}
          onClick={closeDrawer}
        >
          <ListItemIcon>
            <PeopleIcon color="secondary" />
          </ListItemIcon>
          <ListItemText primary={"Partners"} />
        </ListItem>
        <ListItem
          button
          component={RouterLink}
          to="/messages"
          key={"messages"}
          className={classes.listItem}
          onClick={closeDrawer}
        >
          <ListItemIcon>
            <MailIcon color="secondary" />
          </ListItemIcon>
          <ListItemText primary={"Messages"} />
        </ListItem>
        <ListItem
          button
          component={RouterLink}
          to="/profile"
          key={"profile"}
          className={classes.listItem}
          onClick={closeDrawer}
        >
          <ListItemIcon>
            <AccountCircleIcon color="secondary" />
          </ListItemIcon>
          <ListItemText primary={"Profile"} />
        </ListItem>
        <ListItem
          button
          key={"logout"}
          onClick={logout}
          className={classes.listItem}
        >
          <ListItemIcon>
            <ExitToAppIcon color="secondary" />
          </ListItemIcon>
          <ListItemText primary={"Log Out"} />
        </ListItem>
      </List>
    </div>
  );

  const container =
    window !== undefined ? () => window().document.body : undefined;

  return (
    <>
      <div className={classes.root}>
        <CssBaseline />
        <HideOnScroll>
          <AppBar position="fixed" className={classes.appBar}>
            <Toolbar>
              {user && (
                <IconButton
                  color="inherit"
                  aria-label="open drawer"
                  edge="start"
                  onClick={handleDrawerToggle}
                  className={classes.menuButton}
                >
                  <MenuIcon />
                </IconButton>
              )}
              <Typography variant="h6" className={classes.title}>
                <Link
                  component={RouterLink}
                  to="/"
                  color="inherit"
                  underline="none"
                >
                  LangChat
                </Link>
              </Typography>
              {!user && (
                <>
                  <Link
                    component={RouterLink}
                    to="/login"
                    color="inherit"
                    className={classes.link}
                  >
                    Log In
                  </Link>
                  <Link component={RouterLink} to="/signup" color="inherit">
                    Sign Up
                  </Link>
                </>
              )}
            </Toolbar>
          </AppBar>
        </HideOnScroll>
        {user && (
          <nav className={classes.drawer} aria-label="navigation">
            <Hidden mdUp>
              <Drawer
                container={container}
                variant="temporary"
                anchor={theme.direction === "rtl" ? "right" : "left"}
                open={mobileOpen}
                onClose={handleDrawerToggle}
                classes={{
                  paper: classes.drawerPaper,
                }}
                ModalProps={{
                  keepMounted: true, // Better open performance on mobile.
                }}
              >
                {drawer}
              </Drawer>
            </Hidden>
            <Hidden smDown>
              <Drawer
                classes={{
                  paper: classes.drawerPaper,
                }}
                variant="permanent"
                open
              >
                {drawer}
              </Drawer>
            </Hidden>
          </nav>
        )}
        <main className={classes.content}>
          <div className={classes.toolbar} />
          {children}
        </main>
      </div>
      <Footer />
    </>
  );
}

export default NavWrapper;
