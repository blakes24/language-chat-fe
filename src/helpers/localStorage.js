export function setLocalStorage(key, val) {
  window.localStorage.setItem(key, JSON.stringify(val));
}

export function getFromLocalStorage(key, defaultValue) {
  let value;
  try {
    value = JSON.parse(
      window.localStorage.getItem(key) || JSON.stringify(defaultValue)
    );
  } catch (e) {
    console.error(e);
    value = defaultValue;
  }
  return value;
}
