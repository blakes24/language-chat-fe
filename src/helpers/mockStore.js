import configureStore from "redux-mock-store";

const middlewares = [];
const mockStore = configureStore(middlewares);

const testUser = {
  id: 1,
  name: "Test",
  bio: "stuff about mae",
  email: "test@mail.com",
  imageUrl: "https://cdn.fakercloud.com/avatars/happypeter1983_128.jpg",
  active: true,
  speaks: [
    {
      code: "it",
      language: "Italian",
    },
  ],
  learning: [
    {
      code: "ko",
      language: "Korean",
      level: "2",
    },
  ],
};

const initialState = {
  users: {
    current: testUser,
    items: [],
    token: "",
    loading: "idle",
    error: null,
  },
  rooms: { items: [], current: null, loading: "idle", error: null },
  messages: { items: [], loading: "idle", error: null },
  partners: { items: [], loading: "idle", error: null },
};

export const store = mockStore(initialState);
