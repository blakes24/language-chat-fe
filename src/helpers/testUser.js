import UserContext from "./UserContext";

const testUser = {
  id: 1,
  name: "Test",
  bio: "stuff about mae",
  email: "test@mail.com",
  imageUrl: "https://cdn.fakercloud.com/avatars/happypeter1983_128.jpg",
  active: true,
  speaks: [
    {
      code: "it",
      language: "Italian",
    },
  ],
  learning: [
    {
      code: "ko",
      language: "Korean",
      level: "2",
    },
  ],
};

const UserProvider = ({
  children,
  user = testUser,
  setToken = () => console.log("set token"),
}) => <UserContext.Provider value={{ user }}>{children}</UserContext.Provider>;

export { UserProvider };
