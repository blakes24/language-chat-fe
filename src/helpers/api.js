import axios from "axios";
import { BASE_URL } from "../config";
import { getFromLocalStorage } from "./localStorage";

/** API Class: methods used to connect to the API. */

let token = getFromLocalStorage("token", "");

class ChatApi {
  static token = token;

  static async request(endpoint, data = {}, method = "get") {
    console.debug("API Call:", endpoint, data, method);
    const url = `${BASE_URL}/${endpoint}`;
    const headers = { Authorization: `Bearer ${ChatApi.token}` };
    const params = method === "get" ? data : {};
    try {
      return (await axios({ url, method, data, params, headers })).data;
    } catch (err) {
      let message = err.response
        ? err.response.data.error.message
        : "Server error: try again later";
      throw Array.isArray(message) ? message : [message];
    }
  }

  /** Create a new user and get auth token. */

  static async register(userData) {
    let res = await this.request(`auth/register`, userData, "post");
    this.token = res.token;
    return res.token;
  }

  /** Get user details. */

  static async getUser(userId) {
    let res = await this.request(`users/${userId}`);
    return res.user;
  }

  /** Verifies that the social token provided is valid */

  static async verify(token) {
    let res = await this.request(`auth/validate`, token, "post");
    return res;
  }

  /** Get gets token to authenticate future requests to api */

  static async getToken(data) {
    let res = data.email
      ? await this.request(`auth/token`, data, "post")
      : await this.request(`auth/social-token`, data, "post");
    this.token = res.token;
    return res.token;
  }

  /** Get all users, filter by speaks or learning language */

  static async getAllUsers(filters) {
    let res = await this.request(`users`, filters);
    return res.users;
  }

  /** Get a users chat rooms, filter by chat partner */

  static async getRooms({ userId, partnerId }) {
    let queryParams = { partner: partnerId };
    let res = await this.request(`users/${userId}/rooms`, queryParams);
    return res;
  }

  /** Creates a chat room for two users data = {user1, user2} */

  static async createRoom(data) {
    let res = await this.request(`rooms`, data, "post");
    return res;
  }

  /** Gets messages, filter by room */

  static async getMessages(roomId) {
    let queryParams = { room: roomId };
    let res = await this.request(`messages`, queryParams);
    return res.messages;
  }

  /** Update user info. */

  static async updateUser(id, data) {
    let res = await this.request(`users/${id}`, data, "patch");
    return res.user;
  }

  /** Update language(speaks). */

  static async updateSpeaks(id, data) {
    let res = await this.request(`languages/speaks/${id}`, data, "patch");
    return res.language;
  }

  /** Update language(learning). */

  static async updateLearning(id, data) {
    let res = await this.request(`languages/learning/${id}`, data, "patch");
    return res.language;
  }

  /** Get a users partners */

  static async getPartners(userId) {
    let res = await this.request(`users/${userId}/partners`);
    return res.partners;
  }

  /** Adds a partner */

  static async addPartner(userId, data) {
    let res = await this.request(`users/${userId}/partners`, data, "post");
    return res;
  }

  /** Deletes a partner */

  static async deletePartner({ userId, partnerId }, data) {
    let res = await this.request(
      `users/${userId}/partners/${partnerId}`,
      data,
      "delete"
    );
    return res;
  }

  /** Deletes a user */

  static async deleteUser(userId, data) {
    let res = await this.request(`users/${userId}`, data, "delete");
    return res;
  }
}

export default ChatApi;
