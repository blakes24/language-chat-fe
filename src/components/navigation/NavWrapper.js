import React, { useState } from "react";
import AppBar from "@material-ui/core/AppBar";
import CssBaseline from "@material-ui/core/CssBaseline";
import Divider from "@material-ui/core/Divider";
import Drawer from "@material-ui/core/Drawer";
import Hidden from "@material-ui/core/Hidden";
import IconButton from "@material-ui/core/IconButton";
import List from "@material-ui/core/List";
import ListItem from "@material-ui/core/ListItem";
import ListItemIcon from "@material-ui/core/ListItemIcon";
import ListItemText from "@material-ui/core/ListItemText";
import MenuIcon from "@material-ui/icons/Menu";
import Toolbar from "@material-ui/core/Toolbar";
import Typography from "@material-ui/core/Typography";
import { makeStyles, useTheme } from "@material-ui/core/styles";
import Link from "@material-ui/core/Link";
import {
  Link as RouterLink,
  useRouteMatch,
  useHistory,
} from "react-router-dom";
import HomeIcon from "@material-ui/icons/Home";
import PeopleIcon from "@material-ui/icons/People";
import AccountCircleIcon from "@material-ui/icons/AccountCircle";
import ExitToAppIcon from "@material-ui/icons/ExitToApp";
import ChatIcon from "@material-ui/icons/Chat";
import useScrollTrigger from "@material-ui/core/useScrollTrigger";
import Slide from "@material-ui/core/Slide";
import Footer from "./Footer";
import { useDispatch, useSelector } from "react-redux";
import { setToken } from "../../store/usersSlice";
import { setLocalStorage } from "../../helpers/localStorage";

const useStyles = makeStyles((theme) => ({
  root: {
    display: "flex",
    paddingBottom: "env(safe-area-inset-bottom)",
  },
  drawer: {
    [theme.breakpoints.up("md")]: {
      width: 220,
      flexShrink: 0,
    },
  },
  appBar: {
    zIndex: theme.zIndex.drawer + 1,
  },
  menuButton: {
    marginRight: theme.spacing(2),
    [theme.breakpoints.up("md")]: {
      display: "none",
    },
  },
  // necessary for content to be below app bar
  toolbar: theme.mixins.toolbar,
  drawerPaper: {
    width: 220,
    backgroundColor: theme.palette.secondary.light,
  },
  content: {
    flexGrow: 1,
    minHeight: "100vh",
  },
  link: {
    marginLeft: theme.spacing(2),
    fontSize: "1rem",
  },
  title: {
    flexGrow: 1,
  },
  list: {
    margin: "1rem",
  },
  listItem: {
    marginTop: "1rem",
  },
}));

function HideOnScroll({ children }) {
  const trigger = useScrollTrigger();

  return (
    <Slide appear={false} direction="down" in={!trigger}>
      {children}
    </Slide>
  );
}

function NavWrapper({ window, children }) {
  const classes = useStyles();
  const theme = useTheme();
  const [mobileOpen, setMobileOpen] = useState(false);
  const user = useSelector((state) => state.users.current);
  const dispatch = useDispatch();
  const history = useHistory();

  let match = useRouteMatch("/chats");

  function logout() {
    setMobileOpen(false);
    setLocalStorage("token", "");
    dispatch(setToken(""));
    history.push("/");
  }

  const handleDrawerToggle = () => {
    setMobileOpen(!mobileOpen);
  };

  const closeDrawer = () => {
    setMobileOpen(false);
  };

  const drawer = (
    <div>
      <div className={classes.toolbar} />
      <Divider />
      <List className={classes.list}>
        <ListItem
          button
          component={RouterLink}
          to="/"
          key={"home"}
          className={classes.listItem}
          onClick={closeDrawer}
        >
          <ListItemIcon>
            <HomeIcon color="secondary" />
          </ListItemIcon>
          <ListItemText primary={"Home"} />
        </ListItem>
        <ListItem
          button
          component={RouterLink}
          to="/partners"
          key={"partners"}
          className={classes.listItem}
          onClick={closeDrawer}
        >
          <ListItemIcon>
            <PeopleIcon color="secondary" />
          </ListItemIcon>
          <ListItemText primary={"Partners"} />
        </ListItem>
        <ListItem
          button
          component={RouterLink}
          to="/chats"
          key={"chats"}
          className={classes.listItem}
          onClick={closeDrawer}
        >
          <ListItemIcon>
            <ChatIcon color="secondary" />
          </ListItemIcon>
          <ListItemText primary={"Chats"} />
        </ListItem>
        <ListItem
          button
          component={RouterLink}
          to="/profile"
          key={"profile"}
          className={classes.listItem}
          onClick={closeDrawer}
        >
          <ListItemIcon>
            <AccountCircleIcon color="secondary" />
          </ListItemIcon>
          <ListItemText primary={"Profile"} />
        </ListItem>
        <ListItem
          button
          key={"logout"}
          onClick={logout}
          className={classes.listItem}
        >
          <ListItemIcon>
            <ExitToAppIcon color="secondary" />
          </ListItemIcon>
          <ListItemText primary={"Log Out"} />
        </ListItem>
      </List>
    </div>
  );

  const container =
    window !== undefined ? () => window().document.body : undefined;

  return (
    <>
      <div className={classes.root}>
        <CssBaseline />
        <HideOnScroll>
          <AppBar position="fixed" elevation={0} className={classes.appBar}>
            <Toolbar>
              {user && (
                <IconButton
                  color="inherit"
                  aria-label="open drawer"
                  edge="start"
                  onClick={handleDrawerToggle}
                  className={classes.menuButton}
                >
                  <MenuIcon />
                </IconButton>
              )}
              <Typography variant="h6" className={classes.title}>
                <Link
                  component={RouterLink}
                  to="/"
                  color="inherit"
                  underline="none"
                >
                  LangChat
                </Link>
              </Typography>
              {!user && (
                <>
                  <Link
                    component={RouterLink}
                    to="/login"
                    color="inherit"
                    className={classes.link}
                  >
                    Log In
                  </Link>
                  <Link
                    component={RouterLink}
                    to="/signup"
                    color="inherit"
                    className={classes.link}
                  >
                    Sign Up
                  </Link>
                </>
              )}
            </Toolbar>
          </AppBar>
        </HideOnScroll>
        {user && (
          <nav className={classes.drawer} aria-label="navigation">
            <Hidden mdUp>
              <Drawer
                container={container}
                variant="temporary"
                anchor={theme.direction === "rtl" ? "right" : "left"}
                open={mobileOpen}
                onClose={handleDrawerToggle}
                classes={{
                  paper: classes.drawerPaper,
                }}
                ModalProps={{
                  keepMounted: true, // Better open performance on mobile.
                }}
              >
                {drawer}
              </Drawer>
            </Hidden>
            <Hidden smDown>
              <Drawer
                classes={{
                  paper: classes.drawerPaper,
                }}
                variant="permanent"
                open
              >
                {drawer}
              </Drawer>
            </Hidden>
          </nav>
        )}
        <main className={classes.content}>
          <div className={classes.toolbar} />
          {children}
        </main>
      </div>
      {!match && <Footer />}
    </>
  );
}

export default NavWrapper;
