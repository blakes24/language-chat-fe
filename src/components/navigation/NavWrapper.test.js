import { render } from "@testing-library/react";
import NavWrapper from "./NavWrapper";
import { MemoryRouter } from "react-router";
import { store } from "../../helpers/mockStore";
import { Provider } from "react-redux";

test("renders without crashing", () => {
  render(
    <Provider store={store}>
      <MemoryRouter>
        <NavWrapper />
      </MemoryRouter>
    </Provider>
  );
});

it("matches snapshot", function () {
  const { asFragment } = render(
    <Provider store={store}>
      <MemoryRouter>
        <NavWrapper />
      </MemoryRouter>
    </Provider>
  );
  expect(asFragment()).toMatchSnapshot();
});
