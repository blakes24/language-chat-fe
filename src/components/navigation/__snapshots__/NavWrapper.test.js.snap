// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-root-15"
  >
    <header
      class="MuiPaper-root MuiAppBar-root MuiAppBar-positionFixed MuiAppBar-colorPrimary makeStyles-appBar-17 mui-fixed MuiPaper-elevation0"
    >
      <div
        class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
      >
        <button
          aria-label="open drawer"
          class="MuiButtonBase-root MuiIconButton-root makeStyles-menuButton-18 MuiIconButton-colorInherit MuiIconButton-edgeStart"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <h6
          class="MuiTypography-root makeStyles-title-23 MuiTypography-h6"
        >
          <a
            class="MuiTypography-root MuiLink-root MuiLink-underlineNone MuiTypography-colorInherit"
            href="/"
          >
            LangChat
          </a>
        </h6>
        <a
          class="MuiTypography-root MuiLink-root MuiLink-underlineHover makeStyles-link-22 MuiTypography-colorInherit"
          href="/profile"
        >
          Test
        </a>
        <a
          aria-label="view profile"
          class="MuiAvatar-root MuiAvatar-circle makeStyles-avatar-26"
          href="/profile"
          role="link"
        >
          <img
            alt="Test"
            class="MuiAvatar-img"
            src="https://cdn.fakercloud.com/avatars/happypeter1983_128.jpg"
          />
        </a>
      </div>
    </header>
    <nav
      aria-label="navigation"
      class="makeStyles-drawer-16"
    />
    <main
      class="makeStyles-content-21"
    >
      <div
        class="makeStyles-toolbar-19"
      />
    </main>
  </div>
  <footer
    class="makeStyles-root-27"
  >
    <div>
      <p>
        Â© LangChat
      </p>
      <a
        class="makeStyles-link-28"
        href="http://www.freepik.com"
      >
        Graphics designed by stories/Freepik
      </a>
    </div>
  </footer>
</DocumentFragment>
`;
